### PR TITLE
Propagate Debugger… attributes from container of async/iterator method to the corresponding state machine type.

### LINKED ISSUE #4521: Debugger* attributes not applied to async state machine class
`DebuggerNonUserCode` and `DebuggerStepThrough` attributes are not correctly applied to async state machine type when the attribute was on the enclosing type.

This was noted by @thomaslevesque in https://github.com/dotnet/roslyn/issues/807#issuecomment-130472253

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #4521.

@gafter, @VSadov, @jaredpar, @agocke, @TyOverby Please review.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
Ping. @gafter, @VSadov, @jaredpar, @agocke, @TyOverby Please review.

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs [hunk 1]
```diff
@@ -57,5 +59,37 @@ IMethodSymbol ISynthesizedMethodBodyImplementationSymbol.Method
         {
             get { return KickoffMethod; }
         }
+
+        public sealed override ImmutableArray<CSharpAttributeData> GetAttributes()
+        {
+            if (_attributes.IsDefault)
+            {
+                Debug.Assert(base.GetAttributes().Length == 0);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs]: Shouldn't this be true even if the `if` condition is false?

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs [hunk 2]
```diff
@@ -57,5 +59,37 @@ IMethodSymbol ISynthesizedMethodBodyImplementationSymbol.Method
...
+                // Inherit some attributes from the container of the kickoff method
+                var kickoffType = KickoffMethod.ContainingType;
+                foreach (var attribute in kickoffType.GetAttributes())
+                {
+                    if (attribute.IsTargetAttribute(kickoffType, AttributeDescription.DebuggerNonUserCodeAttribute) ||
+                        attribute.IsTargetAttribute(kickoffType, AttributeDescription.DebuggerStepThroughAttribute))
+                    {
+                        if (builder == null)
+                        {
+                            builder = ArrayBuilder<CSharpAttributeData>.GetInstance(4); // only 4 different attributes are inherited at the moment
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs]: The number here is 4 but the conditional above only uses 2 attributes.  Am I missing something?

**Comment by jaredpar** [comment]
:+1:

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs]: It should, but there is no point to verify it more than once.

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineTypeSymbol.cs]: Good catch, it should be changed to 2.

**Pull request merged**